crates/bevy_ecs/src/query/state.rs

```diff
@@ -308,6 +334,36 @@ impl<D: QueryData, F: QueryFilter> QueryState<D, F> {
         }
     }
 
+    fn update_archetypes_with(&mut self, world: &World, f: impl FnMut(&mut Self, &Archetype)) {
+        self.update_archetypes_unsafe_world_cell_with(world.as_unsafe_world_cell_readonly(), f);
+    }
+
+    fn update_archetypes_unsafe_world_cell_with(
+        &mut self,
+        world: UnsafeWorldCell,
+        mut f: impl FnMut(&mut Self, &Archetype),
+    ) {
+        self.validate_world(world.id());
+        if self.component_access.required.is_empty() {
+            let old_generation = std::mem::replace(
+                &mut self.archetype_generation,
+                world.archetypes().generation(),
+            );
+            for archetype in &world.archetypes()[old_generation..] {
+                f(self, archetype);
+            }
+        } else {
+            let mut archetypes = self.get_potential_archetypes(world.archetypes());
```

Comment:
The allocation of the vector and the sort likely make this slower than the original implementation. In order to make this fast we would need to keep an index into the list of archetypes in the component index to continue from which I don't think is feasible (especially since the ordering isn't currently maintained there).

The "update" case should probably stay the way it is for the moment, that will be sped up/made redundant with the introduction of observers and queries as entities. Only the initial population of the cache can realistically be sped up now. I don't believe that is currently special cased so you will need to add a check for the initial archetype generation and use that to opt in to the fast path, ideally we just separate those two code paths altogether.

Reply:
Yes I'll remove the vector allocation + sort; I just added them for convenience to get to an initial working implementation.

You're saying that the "update" case should stay the way it is because we might iterate through previously considered archetypes again?
(even if it's just to filter them with  `.filter(|id| **id >= self.archetype_generation.0)`)
It's true that ComponentIndex might not bring much speed benefits until we have a lot of components...

I can separate the `new()` vs `update()` usecases

Reply:
Yes, perhaps checking the new archetypes against the ones in the component index is fast enough to show an improvement after the allocation and sort are removed, but we would need benchmarks for that.

Reply:
I ran benchmarks on iteration, the PR seems relatively neutral. (but maybe we need better benchmarks? I don't think we have a benchmark that checks iteration separately from query creation)